manticore/native/models.py

```diff
@@ -116,33 +116,41 @@ def strcmp(state: State, s1: Union[int, BitVec], s2: Union[int, BitVec]):
     return ret
 
 
-def strlen(state: State, s: Union[int, BitVec]) -> Union[int, BitVec]:
+def strlen(state: State, addr: Union[int, BitVec]) -> Union[int, BitVec]:
     """
-    strlen symbolic model.
+    strlen symbolic model
 
-    Algorithm: Walks from end of string not including NULL building ITE tree when current byte is symbolic.
+    Algorithm: Counts the number of characters in a string forking every time a symbolic byte
+    is found that can be NULL but is not constrained to NULL.
 
-    :param State state: current program state
+    :param state: current program state
     :param s: Address of string
     :return: Symbolic strlen result
-    :rtype: Expression or int
     """
 
-    cpu = state.cpu
+    if issymbolic(addr):
+        raise ConcretizeArgument(state.cpu, 0)
 
-    if issymbolic(s):
-        raise ConcretizeArgument(state.cpu, 1)
+    cpu = state.cpu
+    constrs = state.constraints
 
-    zero_idx = _find_zero(cpu, state.constraints, s)
+    # Initialize offset based on whether state has been forked in strlen
+    if "strlen" not in state.context:
+        offset = 0
+    else:
+        offset = state.context["strlen"]
 
-    ret = zero_idx
+    c = cpu.read_int(addr + offset, 8)
+    while not is_definitely_NULL(c, constrs):
+        # If the byte can be NULL concretize and fork states
+        if can_be_NULL(c, constrs):
+            state.context["strlen"] = offset
+            raise Concretize("Forking on possible NULL strlen", expression=(c == 0), policy="ALL")
```

Comment:
If you don't want to risk aggressively forking on every single byte, you can sometimes build up a big bitvec If-Then-Else that will accurately capture all the possible values you could concretize to. You can see a similar technique in [ARM's CLZ](https://github.com/trailofbits/manticore/blob/master/manticore/native/cpu/arm.py#L1375-L1387), and also in the previous implementation of `strlen` right here. There are pros and cons to each approach, and I notice that the old implementation used `_find_zero`, which stops at the first byte that _can_ be null, not the first byte that _must_ be null as yours does. Are there any points of interest you can share that drove you toward one approach or the other?